File: R/ideogram.R
# README
# This script creates a linear chromosome plot from structural variant data using ggplot2.
# The script performs the following steps:
# 1. Installs and loads the necessary packages.
# 2. Reads the structural variant data from a file.
# 3. Cleans and processes the data.
# 4. Creates a linear chromosome plot using ggplot2.

# Author: Bloodmark

# Input details:
# - Structural variant data file: "/home/bloodmark/workarea/20231025/LG5/align_out_indels.txt"
#   - The file should be tab-delimited with columns: Scaffold, Type, start, end, and Size.

# Output details:
# - A linear chromosome plot created using ggplot2.

# Script functioning:
# 1. Install and load the necessary packages (if not already installed).
# 2. Read the structural variant data from the input file.
# 3. Clean and process the data by removing rows with NA values and ensuring Size is greater than or equal to 1.
# 4. Create a linear chromosome plot using ggplot2 with custom color palette and size scale.

# Example of running the script:
# Rscript your_script_name.R

# Note: Ensure the input file is in the specified path and properly formatted.

# Code starts below:

# Install required packages if not already installed
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("karyoploteR")
BiocManager::install("Biostrings")

install.packages("ggplot2")
install.packages("colorspace")
install.packages("circlize")
install.packages("RColorBrewer")

# Load required libraries
library(karyoploteR)
library(Biostrings)
library(ggplot2)
library(colorspace)
library(circlize)
library(RColorBrewer)

# Read structural variant data from file
sv_data <- read.table("/home/bloodmark/workarea/20231025/LG5/align_out_indels.txt", header = FALSE, sep = "\t", fill = TRUE)

# Display the first few rows of the data
head(sv_data)

# Set column names
colnames(sv_data) <- c("Scaffold", "Type", "start", "end", "Size")

# Remove rows where Size is less than 1 and remove rows with NA values
sv_data <- sv_data[sv_data$Size >= 1, ]
sv_data <- na.omit(sv_data)

# Determine the number of unique values in sv_data$Type
num_unique_types <- length(unique(sv_data$Type))

# Choose a custom color palette with dark, easily visible colors
custom_palette <- c("#1f77b4", "#ff7f0e", "#2ca02c", "#d62728", "#9467bd", "#8c564b", "#e377c2", "#7f7f7f")

# Create a linear chromosome plot
chromosome_plot <- ggplot(sv_data, aes(x = start, xend = end, y = Type, color = Type)) +
  geom_segment(aes(yend = Type, size = Size)) +
  scale_color_manual(values = custom_palette) +
  scale_size_continuous(range = c(1, 10)) +
  theme_minimal() +
  labs(x = "Position on Chromosome", y = "Feature Type") +
  scale_y_discrete(limits = unique(sv_data$Type))  # Set a common y-axis scale

# Display the plot
print(chromosome_plot)

File: python/ideogram.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Oct 24 20:25:36 2023

@author: bloodmark
"""

# README
# This script reads a PAF file, extracts relevant fields, determines the type of mapping, calculates the size of the mapped region, and writes the information to a BED file.
# The script performs the following steps:
# 1. Reads the PAF file line by line.
# 2. Skips header lines starting with "LN:".
# 3. Extracts relevant fields and checks if the fields should be treated as integers or strings.
# 4. Calculates the size of the mapped region if applicable.
# 5. Determines the type of mapping (insertion, deletion, inversion, or unknown).
# 6. Writes the information to a BED file.

# Author: Bloodmark

# Input details:
# - Input PAF file: "/home/bloodmark/workarea/20231025/chrY.paf"
#   - The file should contain alignment information in PAF format.

# Output details:
# - Output BED file: "output.bed"
#   - The file will contain columns: reference name, start position, end position, query name, size, alignment type.

# Script functioning:
# 1. Reads the PAF file line by line.
# 2. Skips header lines starting with "LN:".
# 3. Extracts relevant fields and checks if the fields should be treated as integers or strings.
# 4. Calculates the size of the mapped region if applicable.
# 5. Determines the type of mapping (insertion, deletion, inversion, or unknown).
# 6. Writes the information to a BED file.

# Example of running the script:
# python your_script_name.py

# Note: Ensure the input file is in the specified path and properly formatted.

with open("/home/bloodmark/workarea/20231025/chrY.paf", "r") as paf_file, open("output.bed", "w") as bed_file:
    for line in paf_file:
        # Skip header lines that start with "LN:"
        if line.startswith("LN:"):
            continue
        
        fields = line.strip().split("\t")
        
        # Check if the line has enough fields (at least 12) before processing
        if len(fields) < 12:
            continue
        
        qname, rname = fields[0], fields[5]
        
        # Check if the fields should be treated as integers or strings
        qstart = int(fields[2]) if fields[2].isdigit() else fields[2]
        qend = int(fields[3]) if fields[3].isdigit() else fields[3]
        rstart = int(fields[7]) if fields[7].isdigit() else fields[7]
        mapping_type = fields[4]
        
        # Calculate size (if applicable)
        if isinstance(qend, int) and isinstance(qstart, int):
            size = qend - qstart
        else:
            size = None
        
        # Determine the type based on mapping type
        if mapping_type == "I":
            alignment_type = "insertion"
        elif mapping_type == "D":
            alignment_type = "deletion"
        elif mapping_type == "N":
            alignment_type = "inversion"
        else:
            alignment_type = "unknown"
        
        # Write the information to the BED file
        if size is not None:
            bed_file.write(f"{rname}\t{rstart}\t{rstart + size}\t{qname}\t{size}\t{alignment_type}\n")
        else:
            bed_file.write(f"{rname}\t{rstart}\t.\t{qname}\t.\t{alignment_type}\n")

